Using access lists

sunce86 opened the pull request:
This PR implements using the access lists for two components, settlement simulation and settlement submission.

sunce86 committed:
eth_createAccessList implementation

sunce86 committed:
renaming

sunce86 committed:
added PriorityAccessListEstimator

sunce86 committed:
add meaningful message

sunce86 committed:
using BatchTransport

sunce86 committed:
small fix

sunce86 committed:
code review fixes

sunce86 committed:
renamed uts

sunce86 committed:
Using access lists

sunce86 committed:
Merge branch 'main' into sunce86/using-access-lists

sunce86 committed:
testing fixes - added fix for empty txs - added calling filter_access_list - fix for `to` address

sunce86 committed:
fix for simulation

sunce86 committed:
substitute batch transport for settlement simulation

sunce86 committed:
Merge branch 'main' into sunce86/using-access-lists

vkgnosis commented on crates/solver/src/settlement_simulation.rs:
```diff
@@ -48,41 +54,47 @@ pub async fn simulate_and_estimate_gas_at_current_block(
+                .tx
+            })
+            .collect::<Vec<_>>();
+        let mut access_lists = access_list_estimator
+            .estimate_access_lists(txs.as_slice())
```
I'm wondering if it would be cleaner to create the access lists when we also create the rest of the transaction itself and then pass it around to the simulation here and to the submitter later. That way the access list only has to be created once. Also it feels more proper to me that the access list would be part of the transaction and not be created on demand in several places.

sunce86 replied:
Access list is something that is theoretically dependent on the current network state, so it makes sense to me to be created at the moment of using. That is the reason why I estimate access list in every iteration of the submission loop, instead of doing it only once before entering the loop. Not sure about this though, if others think we should calculate it only once, I can change it

nlordell replied:
Makes sense to me. It reduces the change of paying for storage access you don't end up using.

sunce86 replied:
Refactored these two functions to receive already prepared access list. Access lists are now created once in driver.

MartinquaXD commented on crates/solver/src/settlement_simulation.rs:
```diff
@@ -48,41 +54,47 @@ pub async fn simulate_and_estimate_gas_at_current_block(
+        let calls = txs
+            .into_iter()
+            .map(|tx| {
+                let tx = match access_lists.next().unwrap() {
```
I believe `unwrap()` is able to panic here if `estimate_access_lists()` doesn't return an access list for every transaction because some bogous acess lists got filtered out.

Also if `estimate_access_lists()` throws an error `unwrap_or_default()` would return an empty `Vec` resulting in the same problem.

sunce86 replied:
Changed the logic a bit.

MartinquaXD commented on crates/solver/src/settlement_simulation.rs:
```diff
@@ -91,12 +103,36 @@ pub async fn simulate_and_error_with_tenderly_link(
     simulation_gas_limit: u128,
+    access_list_estimator: Arc<dyn AccessListEstimating>,
 ) -> Vec<Result<()>> {
+    let settlements = settlements.collect::<Vec<_>>();
```
I think this `collect()` can be dropped by using the `settlements` iterator directly a few lines down.

sunce86 replied:
Yup. Copy paste.

MartinquaXD commented on crates/solver/src/settlement_simulation.rs:
```diff
@@ -91,12 +103,36 @@ pub async fn simulate_and_error_with_tenderly_link(
+    // append access lists to existing methods if access lists exist, otherwise ignore access lists
+    let methods = methods
+        .into_iter()
+        .map(|method| match access_lists.next().unwrap() {
```
Same problem with `unwrap()` as above.

sunce86 replied:
Fixed.

MartinquaXD commented on crates/solver/src/settlement_simulation.rs:
```diff
@@ -91,12 +103,36 @@ pub async fn simulate_and_error_with_tenderly_link(
+            Ok(access_list) => method.access_list(access_list),
+            Err(_) => method,
+        })
+        .collect::<Vec<_>>();
```
No `collect()` needed here.

sunce86 replied:
True :)

sunce86 committed:
added tenderly url and api key

sunce86 committed:
code review fixes

sunce86 committed:
fmt

nlordell commented on crates/solver/src/main.rs:
```diff
@@ -184,6 +185,32 @@ struct Arguments {
+        ignore_case = true,
+        use_value_delimiter = true
+    )]
+    pub access_list_estimators: Vec<AccessListEstimatorType>,
```
ubernit:
```suggestion
    access_list_estimators: Vec<AccessListEstimatorType>,
```

sunce86 replied:
Fixed.

nlordell commented on crates/solver/src/settlement_submission/submitter.rs:
```diff
@@ -436,51 +444,88 @@ impl<'a> Submitter<'a> {
+        &self,
+        tx: &TransactionBuilder<DynTransport>,
+    ) -> Result<AccessList> {
+        let access_list = self.access_list_estimator.estimate_access_list(tx).await?;
```
Maybe the sanitization of the access list should also move here? This way each implementation doesn't need to call it themselves.

sunce86 replied:
Hm, then I would have to add the sanitization for settlement simulation also, so it's not that beneficial, What we can do is to add the sanitization to `PriorityAccessListEstimating` but also make `NodeApi` and `TenderlyApi` private so we are forced to use `PriorityAccessListEstimating`

sunce86 replied:
Fixed as explained.
Also feels like filtering is the response of the estimator, not the client.

nlordell reviewed (commented):
In general looks good.

How long does it take to generate access lists? I'm a bit concerned that we are adding more time to preparing and simulating settlement transactions which may have issues if we get "too busy".

Agree that we should re-compute the access list on each submission loop - but maybe we can try and reuse the first one (also reuse it for tenderly link).

sunce86 committed:
nicks code review

sunce86 commented on crates/solver/src/settlement_submission/submitter.rs:
```diff
@@ -436,51 +444,88 @@ impl<'a> Submitter<'a> {
+        let access_list = self.access_list_estimator.estimate_access_list(tx).await?;
+        let gas_before_access_list = tx.clone().estimate_gas().await?;
+        let tx = tx.clone().access_list(access_list.clone());
+        let gas_after_access_list = tx.clone().estimate_gas().await?;
```
Indeed.

sunce86 committed:
calculating access lists only once in driver

sunce86 commented:
Based on the logs it takes around 130ms for tenderly. Anyway, it shouldn't be an issue, will be checking the logs on staging.

nlordell commented:
> Based on the logs it takes around 130ms for tenderly.

My concern is more about this during "busy times". For example, if we need to simulate 30 settlements (which can happen because of settlement merging for example) then we need to compute 30x access lists.

To be clear, I'm not too worried about this to the point that I think it should block this change, just hoping we could find a way to quantify it.

sunce86 committed:
access list creation moved out of loop

nlordell commented:
Thanks very much for looking into this for me @sunce86.

nlordell commented on crates/solver/src/main.rs:
```diff
@@ -184,6 +185,32 @@ struct Arguments {
+    #[clap(
+        long,
+        env,
+        default_values = &["Web3", "Tenderly"],
```
nit: Can we make the default an empty list. My rationale is:
- I _think_ if you have default values for `use_value_delimiter` parameters, there is actually no way to configure it to use an empty list. At least this was the case a few months ago.
- Including `Tenderly` in the default requires an access key - this is annoying to configure for local testing
- Including `Web3` requires a node with `createAccessList` support which is not always the case for local testing or for all networks (does this work on GChain and Rinkeby for example?)

That being said, I do think that the value for our Mainnet solver pod should be what you have here, just maybe it makes more sense to set that in our Kubernetes deployment configuration instead of hardcoded defaults in the binary.

sunce86 replied:
Yes, makes sense. Removed default.

> Including Web3 requires a node with createAccessList support which is not always the case for local testing or for all networks (does this work on GChain and Rinkeby for example?)

It does require a node with `createAccessList` support, however the code should not panic in case we define `Web3` and `createAccessList` is not supported on that node, because we agreed to have the access list functionality as an additional bonus, and to fallback to no access list if one does not exist for any reason.
Both client node (if it implements `createAccessList`) and tenderly support GChain and Rinkeby.

nlordell reviewed (approved):
Changes look good to me!

One small follow-up: Now that we have access-list support, we should be able to allow Smart Contracts to place orders buying Ether 🚀. So, we can now allow a `--allow-smart-contracts-buying-native-token` in the `orderbook` so that we can turn on this feature (my rationale for not having it on by default is that we can, in theory, configure no access list estimators, which means we need to be able to disable it in the orderbook API as well).

sunce86 committed:
removed default value for access list estimators,
removed requirement to have at least 1 access list estimator in order to start the application

sunce86 committed:
tenderly url made optional without default value

sunce86 commented on crates/solver/src/main.rs:
```diff
@@ -184,6 +185,32 @@ struct Arguments {
+    #[clap(
+        long,
+        env,
+        default_value = "http://api.tenderly.co/api/v1/account/sunce86/project/project/simulate"
```
Indeed. Should probably define a new endpoint for deployment configuration. Fixed.

sunce86 committed:
Merge branch 'main' into sunce86/using-access-lists
